README.md
# chardetection

Character encoding auto-detection in JavaScript (using jschardet and chardet)

[![build status](https://secure.travis-ci.org/fasterize/chardetection.svg)](http://travis-ci.org/fasterize/chardetection)
[![dependency status](https://david-dm.org/fasterize/chardetection.svg)](https://david-dm.org/fasterize/chardetection)

## Motivation

[Jschardet](https://github.com/aadsm/jschardet) is a Javascript port of mozilla universal character detector. It is mainly used by the editor [Vscode](https://github.com/Microsoft/vscode/blob/master/src/vs/base/node/encoding.ts) and [Atom](https://github.com/atom/encoding-selector/blob/master/lib/main.js)
[chardet](https://www.npmjs.com/package/chardet) is a character detector based on ICU. It is mainly used by the discord-irc project.

I found that those two char detectors could be very inacurrate on some situation.
Combining both detectors give better detection results.

This module takes into account the support provided by both detectors and also the [current usage of charsets on the web](https://w3techs.com/technologies/overview/character_encoding/all). In case of doubt, the priority is given to the most used charset.

This module comes with an extensive tests suite compiled from other detectors test suites.

## Algorithm

- Jschardet is used as a primary source
- => detected charset is on the list of accurately detected charset by jschardet, return

- Chardet ICU is used as a second source
- => detected charset is on the list of accurately detected charset by chardet, return

- Verify if both chardet detect the same chardet, return

- Send the jschardet detected charset.

## Benchmark

- chardetector : 81 errors over 540 files
- jschardet : 133 errors over 540 files
- chardet : 147 errors over 540 files
- mmmagic : 434 errors over 540 files

## Installation

```
npm install --save chardetection
```

## Usage

```js
import chardetector from 'chardetector';

const charset = chardetector(
  buffer,
  [debug],
  [logger],
  [bufferMaxLength],
  [charsetsWithLowConfidence],
  [jschardetCharsetList],
  [chardetCharsetList],
);
```

## Credits

[Fasterize](https://github.com/fasterize/)

## License

ISC


CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.1] - 2018-10-12
Don't download the test suite for npm usage

## [1.0.0] - 2018-10-12
First version leveraging jschardet and chardet


test/testhelper.js
/* eslint-env mocha */
const should = require('should');
const iconv = require('iconv-lite');
const fs = require('fs');
const path = require('path');
const assert = require('assert');

const deprecatedCharset = [
  'ISO-2022-jp',
  'ISO-2022-kr',
  'MacCyrillic',
  'IBM855',
  'IBM866',
  'IBM424',
  'UTF-16',
  'UTF-16LE',
  'UTF-16BE',
  'UTF-32',
];

module.exports = function charsetFullSuite(detector) {
  function testEncoding(file, expectedCharset) {
    const encodedBuffer = fs.readFileSync(file);
    const charset = detector(encodedBuffer);
    if (expectedCharset) {
      should.exists(charset);
      iconv.encodingExists(charset).should.be.true();
      assert.equal(
        iconv.decode(encodedBuffer, charset),
        iconv.decode(encodedBuffer, expectedCharset),
        `detected charset ${charset}`,
      );
    } else {
      should.not.exists(charset);
    }
  }

  const charsetDirectory = path.join(__dirname, '/encodings');
  const charsets = fs.readdirSync(charsetDirectory);
  charsets.filter(charset => !deprecatedCharset.includes(charset)).forEach(charset => {
    describe(`charset ${charset} detection`, () => {
      const files = fs.readdirSync(`${charsetDirectory}/${charset}`);
      files.forEach(file => {
        const expectedCharset = charset.replace(
          /-(arabic|bulgarian|cyrillic|greek|hebrew|hungarian|turkish)/,
          '',
        );
        it(`should detect the charset ${expectedCharset} for the file ${charset}/${file}`, () => {
          testEncoding(`${charsetDirectory}/${charset}/${file}`, expectedCharset);
        });
      });
    });
  });

  describe('binary content', () => {
    const directory = path.join(__dirname, 'binary');
    const files = fs.readdirSync(directory);
    files.forEach(file => {
      it(`should not detect a charset for the file ${file}`, () => {
        testEncoding(`${directory}/${file}`, null);
      });
    });
  });
};


test/chardet.text.js
/* eslint-env mocha */
const chardet = require('chardet');
const charsetFullSuite = require('./testhelper');

describe('test charset chardet', () => {
  charsetFullSuite(buffer => chardet.detect(buffer.slice(0, 30 * 1024)));
});


test/mmmagic.text.js
/* eslint-env mocha */
const mmm = require('mmmagic');
const charsetFullSuite = require('./testhelper-async');

const magic = new mmm.Magic(mmm.MAGIC_MIME_ENCODING);

describe('test charset jschardet', () => {
  charsetFullSuite((buffer, callback) => {
    try {
      magic.detect(buffer.slice(0, 30 * 1024), callback);
    } catch (error) {
      console.log(error);
    }
  });
});


index.js
const chardet = require('chardet'); // icu detector
const jschardet = require('jschardet'); // mozilla universal detector

const defaultJschardetCharsetList = [
  'UTF-8',
  'ascii',
  'ISO-8859-5',
  'KOI8-R',
  'TIS-620',
  'SHIFT_JIS',
  'EUC-JP',
  'EUC-KR',
  'GB2312',
  'ISO-2022-JP',
  'ISO-2022-KR',
  'Big5',
];

const defaultChardetCharsetList = [
  'UTF-8',
  'ISO-8859-2',
  'ISO-8859-6',
  'ISO-8859-9',
  'EUC-KR',
  'windows-1251',
  'windows-1252',
  'windows-1255',
  'windows-1254',
  'windows-1256',
];

const defaultBufferMaxLength = 20 * 1024;

const defaultCharsetsWithLowConfidence = [
  'ISO-2022-JP',
  'ISO-2022-KR',
  'MacCyrillic',
  'IBM855',
  'IBM866',
];

const normalizeUTF8 = charset => {
  // reduce risk of wrong detection since ascii is a subset of UTF-8
  return ['ascii'].includes(charset) ? 'UTF-8' : charset;
};

module.exports = (
  buffer,
  debug = false,
  // eslint-disable-next-line
  logger = console.log,
  bufferMaxLength = defaultBufferMaxLength,
  charsetsWithLowConfidence = defaultCharsetsWithLowConfidence,
  jschardetCharsetList = defaultJschardetCharsetList,
  chardetCharsetList = defaultChardetCharsetList,
) => {
  if (buffer.length > 0) {
    const bodyPart = buffer.slice(0, bufferMaxLength);

    const jschardetResult = jschardet.detect(bodyPart);

    if (debug) {
      logger('Trying detection with jschardet', jschardetResult);
    }

    // very good detection for jschardet
    if (
      jschardetCharsetList.includes(jschardetResult.encoding) &&
      jschardetResult.confidence > 0.9 &&
      !charsetsWithLowConfidence.includes(jschardetResult.encoding)
    ) {
      if (debug) {
        logger('Use jschardet detection');
      }
      return normalizeUTF8(jschardetResult.encoding);
    }

    const chardetResult = chardet.detectAll(bodyPart)[0]; // icu port used by chromium

    if (debug) {
      logger('Trying detection with icu chardet', chardetResult);
    }

    // very good detection for chardet that could be wrong for jschardet
    if (
      chardetCharsetList.includes(chardetResult.name) &&
      chardetResult.confidence > 20 &&
      !charsetsWithLowConfidence.includes(chardetResult.name)
    ) {
      if (debug) {
        logger('Use chardet detection');
      }
      return normalizeUTF8(chardetResult.name);
    }

    // heuristic on the detection of 'ISO-8859-1', 'ISO-8859-15, 'ISO-8859-7''
    // heuristic based on encoding occurance from https://w3techs.com/technologies/overview/character_encoding/all
    // In case of doubt, ISO-8859-1 is selected
    if (jschardetResult.encoding === chardetResult.name) {
      if (debug) {
        logger('Both chardet detect the same charset');
      }
      return normalizeUTF8(jschardetResult.encoding);
    }

    // from https://chardet.readthedocs.io/en/latest/supported-encodings.html
    // Greek text encoded as ISO-8859-7 was often mis-reported as ISO-8859-2
    if (
      chardetResult.name === 'ISO-8859-1' &&
      chardetResult.confidence < 30 &&
      jschardetResult.encoding === 'ISO-8859-2' &&
      jschardetResult.confidence >= 0.6
    ) {
      if (debug) {
        logger('heuristic for ISO-8859-7');
      }
      return 'ISO-8859-7';
    }

    if (chardetResult.name === 'ISO-8859-1' && chardetResult.confidence > 50) {
      if (debug) {
        logger('heuristic for ISO-8859-1');
      }
      return 'ISO-8859-1';
    }

    if (debug) {
      logger('default to jschardet detection');
    }
    return normalizeUTF8(jschardetResult.encoding);
  }
  return null;
};
